src/lib/gossip_net/libp2p.ml

```diff
@@ -813,17 +892,33 @@ module Make (Rpc_intf : Mina_base.Rpc_intf.Rpc_interface_intf) :
         (Peer.pretty_list peers) ;
       List.map peers ~f:(fun peer -> query_peer t peer.peer_id rpc query)
 
-    let broadcast_state t state =
+    (* Do not broadcast to the topic from which message was originally received *)
+    let guard_topic ?origin_topic topic f msg =
+      if Option.equal String.equal origin_topic (Some topic) then Deferred.unit
+      else f msg
+
+    (* broadcast to new topics  *)
+    let broadcast_state ?origin_topic t state =
       let pfs = !(t.publish_functions) in
-      pfs.publish_v0 (Message.New_state state)
+      guard_topic ?origin_topic v1_topic_block pfs.publish_v1_block state
+      >>= fun () ->
```

Comment:
Nit: style-wise, I prefer `let%bind () = ... in ...` over `... >>= fun () -> ...`.